_pytest/cacheprovider.py

```diff
@@ -149,7 +149,9 @@ def pytest_sessionfinish(self, session):
         config = self.config
         if config.getvalue("cacheshow") or hasattr(config, "slaveinput"):
             return
-        config.cache.set("cache/lastfailed", self.lastfailed)
+        prev_failed = config.cache.get("cache/lastfailed", None) is not None
+        if (session.items and prev_failed) or self.lastfailed:
```

Comment:
Try `session.testscollected`... this is also [set by xdist](https://github.com/pytest-dev/pytest-xdist/blob/master/xdist/dsession.py#L636) so the master process knows that some tests were collected.